#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-22-x_edit_marks.dpatch by FrankJepsen at vdrportal.de
## http://www.jepsennet.de/vdr/download/opt-22-x_edit_marks.dpatch
##
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Patch adds keys '1', '3' in replay mode to move marks by +/-5secs.
## DP: When in replay mode key '0' sets a mark and stops there.
## DP: 'Green' and 'Yellow' act as 'Prev' and 'Next'.
@DPATCH@
diff -u vdr-1.6.0p/menu.c vdr-1.6.0n/menu.c
--- vdr-1.6.0p/menu.c   2010-01-15 12:55:13.000000000 +0100
+++ vdr-1.6.0n/menu.c   2010-01-15 13:17:01.000000000 +0100
@@ -5078,10 +5078,6 @@
     case kGreen:   SkipSeconds(-60); break;
     case kYellow|k_Repeat:
     case kYellow:  SkipSeconds( 60); break;
-    case k1|k_Repeat:
-    case k1:       SkipSeconds(-20); break;
-    case k3|k_Repeat:
-    case k3:       SkipSeconds( 20); break;
     case kPrev|k_Repeat:
     case kPrev:    if (lastSkipTimeout.TimedOut()) {
                       lastSkipSeconds = REPLAYCONTROLSKIPSECONDS;
@@ -5109,10 +5105,46 @@
                    Stop();
                    return osEnd;
     default: {
+      bool play, forward;
+      int speed;
+      int Current, Total;
+      cMark *m;
+
       DoShowMode = false;
       switch (Key) {
         // Editing:
-        case kMarkToggle:      MarkToggle(); break;
+        case k1|k_Repeat:
+        case k1:            if (GetReplayMode(play, forward, speed) && !play && GetIndex(Current, Total) && (m = marks.Get(Current)) != NULL) {
+                             displayFrames = true;
+                             int p = SkipFrames(-5 * FRAMESPERSEC);
+                             cMark *m2;
+                             if ((m2 = marks.Prev(m)) != NULL && m2->position >= p)
+                               break;
+                             Goto(m->position = p, true);
+                             marks.Save();
+                           }
+                           else
+                               SkipSeconds(-20);
+                           break;
+        case k3|k_Repeat:
+        case k3:            if (GetReplayMode(play, forward, speed) && !play && GetIndex(Current, Total) && (m = marks.Get(Current)) != NULL) {
+                             displayFrames = true;
+                             int p = SkipFrames(5 * FRAMESPERSEC);
+                             cMark *m2;
+                             if ((m2 = marks.Next(m)) != NULL && m2->position <= p)
+                               break;
+                             Goto(m->position = p, true);
+                             marks.Save();
+                           }
+                           else
+                               SkipSeconds(20);
+                           break;
+        case kMarkToggle:   if (GetReplayMode(play, forward, speed) && play) {
+                              displayFrames = true;
+                              Pause();
+                           }
+                            MarkToggle();
+                            break;
         case kMarkJumpBack|k_Repeat:
         case kMarkJumpBack:    MarkJump(false); break;
         case kMarkJumpForward|k_Repeat:
